algebird-test/src/test/scala/com/twitter/algebird/ScanTest.scala

```diff
@@ -0,0 +1,146 @@
+package com.twitter.algebird
+
+import org.scalacheck.Gen
+import org.scalatestplus.scalacheck.ScalaCheckDrivenPropertyChecks
+import org.scalatest.{Matchers, WordSpec}
+
+import scala.collection.mutable.Queue
+
+object ScanTest {
+  // normal people will use Scan not Scan.Aux, so it's good for most of the tests to be using the more common interface.
+  type StringScan = Scan[Char, String]
+
+  // technically speaking, these aren't exactly the Free scanner, since that would output a giant tree structure from
+  // the whole scan, but that giant tree structure is pretty close to a String.
+
+  val directFreeScan: Scan.Aux[Char, List[Char], String] = Scan.from(List.empty[Char]) {
+    (char, previousState) =>
+      val nextState = char :: previousState
+      (nextState.reverse.mkString, nextState)
+  }
+
+  val mutableFreeScan: StringScan = Scan.mutable(new Queue[Char]()) { (char, previousState) =>
+    previousState.enqueue(char)
+    previousState.mkString
+  }
+
+  val aggregatorFreeScan: StringScan = {
+    val aggregator = Aggregator.fromMonoid[List[Char]]
+
+    Scan
+      .fromMonoidAggregator(aggregator)
+      .composePrepare[Char](c => List(c))
+      .andThenPresent(_.mkString)
+
+  }
+
+  val joinWithPosteriorStateFreeScan: StringScan =
+    directFreeScan
+      .andThenPresent(_ => ())
+      .joinWithPosteriorState
+      .andThenPresent { case ((), state) => state.reverse.mkString }
+
+  val joinWithPriorStateFreeScan1: StringScan =
+    directFreeScan
+      .andThenPresent(_ => ())
+      .joinWithPriorState
+      .joinWithInput
+      .andThenPresent { case (input, (state, ())) => (input :: state).mkString.reverse }
+
+  val joinWithPriorStateFreeScan2: StringScan =
+    directFreeScan
+      .andThenPresent(_ => ())
+      .joinWithPriorState
+      .join(Scan.identity[Char])
+      .andThenPresent { case ((state, ()), input) => (input :: state).mkString.reverse }
```

Comment:
Looking at this again, `joinWithInput` and `join(Scan.identity)` end up putting the joined values in opposite orders. Is it worth bikeshedding which order should be the right one?

Reply:
yes. `WithIndex` in scala puts the Index *after* the value `A => (A, Int)`, so, `WithInput` might also put it on the right to be consistent with that approach.

Particularly dangerous when they are the same type!